File: exercises/javascript/9_substitution/evaluate.js
import {
  emailRegex,
  emailReplacement,
  htmlTagRegex,
  htmlTagReplacement,
  multipleWhitespaceRegex,
  multipleWhitespaceReplacement,
  outerWhitespaceRegex,
  outerWhitespaceReplacement,
  nameSwitchRegex,
  nameSwitchReplacement,
  multiNameSwitchRegex,
  multiNameSwitchReplacement,
} from '.';

describe('EvaluateEmailReplacement', function () {
  it('redacts whole string email', function () {
    const string = 'not.an.email@fakedomain.com';
    const output = string.replace(emailRegex, emailReplacement);
    expect(output).toBe('<<redacted>>');
  });
  it('redacts multiple_emails', function () {
    const string = 'Please send to a@b.com and c@d.com. Thanks!';
    const output = string.replace(emailRegex, emailReplacement);
    expect(output).toBe(
      'Please send to <<redacted>> and <<redacted>>. Thanks!'
    );
  });
  it('does not change string with no emails', function () {
    const string = 'This string has no emails!!';
    const output = string.replace(emailRegex, emailReplacement);
    expect(output).toBe('This string has no emails!!');
  });
});
describe('EvaluateHtmlReplacement', function () {
  const string = it('removes tags for single element', function () {
    const string = '<span>This is a span</span>';
    const output = string.replace(htmlTagRegex, htmlTagReplacement);
    expect(output).toBe('This is a span');
  });
  it('removes tags for multiple elements', function () {
    const string =
      '<span>This is a <b>span!</b></span><span>So is this.</span>';
    const output = string.replace(htmlTagRegex, htmlTagReplacement);
    expect(output).toBe('This is a span!So is this.');
  });
  it('multiline', function () {
    const string =
      '<h1>Regular Expressions</h1>\n<h2>Quantifiers</h2>\n<p>Quantifiers tell you how many of the preceding token are allowed.</p>';
    const output = string.replace(htmlTagRegex, htmlTagReplacement);
    expect(output).toBe(
      'Regular Expressions\nQuantifiers\nQuantifiers tell you how many of the preceding token are allowed.'
    );
  });
  it('does not change string with no tags', function () {
    const string = 'No tags. Absolutely none.';
    const output = string.replace(htmlTagRegex, htmlTagReplacement);
    expect(output).toBe('No tags. Absolutely none.');
  });
});
describe('EvaluateMultipleWhitespaceReplacement', function () {
  const string = 'hello,\t   my   name is  Bonnie';
  it('replaces multiple whitespac for a single-line string', function () {
    const output = string.replace(
      multipleWhitespaceRegex,
      multipleWhitespaceReplacement
    );
    expect(output).toBe('hello, my name is Bonnie');
  });
  it('multi_line', function () {
    const string =
      "it's   the\n\n      most   wonderful\t\t\t time\n      of the\tyear";
    const output = string.replace(
      multipleWhitespaceRegex,
      multipleWhitespaceReplacement
    );
    expect(output).toBe("it's the most wonderful time of the\tyear");
  });
  it('does not change string with no whitespace', function () {
    const string = 'no_multipleWhitespace__at____all';
    const output = string.replace(
      multipleWhitespaceRegex,
      multipleWhitespaceReplacement
    );

    expect(output).toBe('no_multipleWhitespace__at____all');
  });
});
describe('EvaluateOuterWhitespaceReplacement', function () {
  it('removes leading and trailing space', function () {
    const string = '   leading and trailing    \n\n';
    const output = string.replace(
      outerWhitespaceRegex,
      outerWhitespaceReplacement
    );
    expect(output).toBe('leading and trailing');
  });
  it('removes trailing space when preceded with punctuation', function () {
    const string = 'only trailing space!   \t';
    const output = string.replace(
      outerWhitespaceRegex,
      outerWhitespaceReplacement
    );
    expect(output).toBe('only trailing space!');
  });
  it('removes leading space when followed by punctuation', function () {
    const string = '\n"Get to the back of the ship!" Tom said sternly.\n';
    const output = string.replace(
      outerWhitespaceRegex,
      outerWhitespaceReplacement
    );
    expect(output).toBe('"Get to the back of the ship!" Tom said sternly.');
  });
  it('does not change string with no outer space', function () {
    const string = 'Do not launch me out of the atmosphere please!';
    const output = string.replace(
      outerWhitespaceRegex,
      outerWhitespaceReplacement
    );
    expect(output).toBe('Do not launch me out of the atmosphere please!');
  });
});
describe('EvaluateNameSwitchReplacement', function () {
  it('reverses names from the instructions example', function () {
    const string =
      "Indira Ghandi\nMorgan Freeman\nJustin Finch-Fletchley\nFlannery O'Connor";
    const output = string.replace(nameSwitchRegex, nameSwitchReplacement);
    expect(output).toBe(
      "Ghandi, Indira\nFreeman, Morgan\nFinch-Fletchley, Justin\nO'Connor, Flannery"
    );
  });
  it('reverses single single_name', function () {
    const string = 'Ghandi, Indira';
    const output = string.replace(nameSwitchRegex, nameSwitchReplacement);
    expect(output).toBe('Ghandi, Indira');
  });
  it('does not reverse single-word name', function () {
    const output = 'Pele'.replace(nameSwitchRegex, nameSwitchReplacement);
    expect(output).toBe('Pele');
  });
});
describe('EvaluateMultiNameSwitchReplacement', function () {
  it('reverses names from instructionsexample', function () {
    const string =
      "Indira Ghandi\nMorgan Freeman\nJustin Finch-Fletchley\nFlannery O'Connor\nCarlos Alberto Torres\nHarriet Beecher Stowe";
    const output = string.replace(
      multiNameSwitchRegex,
      multiNameSwitchReplacement
    );
    expect(output).toBe(
      "Ghandi, Indira\nFreeman, Morgan\nFinch-Fletchley, Justin\nO'Connor, Flannery\nTorres, Carlos Alberto\nStowe, Harriet Beecher"
    );
  });
  it('reverses single single_name', function () {
    const string = 'Ghandi, Indira';
    const output = string.replace(
      multiNameSwitchRegex,
      multiNameSwitchReplacement
    );
    expect(output).toBe('Ghandi, Indira');
  });
  it('does not reverse single-word name', function () {
    const output = 'Pele'.replace(
      multiNameSwitchRegex,
      multiNameSwitchReplacement
    );
    expect(output).toBe('Pele');
  });
});

File: exercises/python/9_substitution/evaluate.py
from unittest import TestCase
import re

from substitution import email_regex
from substitution import email_replacement
from substitution import html_tag_regex
from substitution import html_tag_replacement
from substitution import multiple_whitespace_regex
from substitution import multiple_whitespace_replacement
from substitution import outer_whitespace_regex
from substitution import outer_whitespace_replacement
from substitution import name_switch_regex
from substitution import name_switch_replacement
from substitution import multi_name_switch_regex
from substitution import multi_name_switch_replacement


def output_does_not_match(regex_name, input_string, expected, actual):
    return f'{regex_name} applied to "{input_string}" produced "{actual}" when it should have been "{expected}".'


class EvaluateEmailReplacement(TestCase):
    regex_name = 'email_regex'

    def test_whole_string_email(self):
        input_string = 'not.an.email@fakedomain.com'
        expected_output = '<<redacted>>'

        output = re.sub(
            email_regex,
            email_replacement,
            input_string,
        )
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output))

    def test_multiple_emails(self):
        input_string = 'Please send to a@b.com and c@d.com. Thanks!'
        expected_output = 'Please send to <<redacted>> and <<redacted>>. Thanks!'
        output = re.sub(
            email_regex,
            email_replacement,
            input_string
        )
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output))

    def test_no_emails(self):

        input_string = 'This string has no emails!!'
        expected_output = 'This string has no emails!!'
        output = re.sub(
            email_regex,
            email_replacement,
            input_string
        )
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output))


class EvaluateHtmlReplacement(TestCase):
    regex_name = 'html_tag_regex'

    def test_simple(self):
        input_string = '<span>This is a span</span>'
        expected_output = 'This is a span'
        output = re.sub(
            html_tag_regex,
            html_tag_replacement,
            input_string
        )
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output)
        )

    def test_multiple(self):
        input_string = '<span>This is a <b>span!</b></span><span>So is this.</span>'
        expected_output = 'This is a span!So is this.'
        output = re.sub(
            html_tag_regex,
            html_tag_replacement,
            input_string
        )
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output)
        )

    def test_multiline(self):
        html = """<h1>Regular Expressions</h1>
<h2>Quantifiers</h2>
<p>Quantifiers tell you how many of the preceding token are allowed.</p>
"""
        output = re.sub(
            html_tag_regex,
            html_tag_replacement,
            html
        )
        self.assertEqual(
            output,
            """Regular Expressions
Quantifiers
Quantifiers tell you how many of the preceding token are allowed.
"""
        )

    def test_no_tags(self):
        input_string = 'No tags. Absolutely none.'
        expected_output = 'No tags. Absolutely none.'
        output = re.sub(
            html_tag_regex,
            html_tag_replacement,
            input_string
        )
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output)
        )


class EvaluateMultipleWhitespaceReplacement(TestCase):
    regex_name = 'multiple_whitespace_regex'

    def test_single_line(self):
        input_string = 'hello,\t   my   name is  Bonnie'
        expected_output = 'hello, my name is Bonnie'
        output = re.sub(
            multiple_whitespace_regex,
            multiple_whitespace_replacement,
            input_string
        )
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output))

    def test_multi_line(self):
        output = re.sub(
            multiple_whitespace_regex,
            multiple_whitespace_replacement,
            """it's   the

        most   wonderful\t\t\t time
        of the\tyear"""
        )
        self.assertEqual(
            output,
            'it\'s the most wonderful time of the\tyear'
        )

    def test_no_whitespace(self):
        input_string = """no_multiple_whitespace__at____all"""
        expected_output = """no_multiple_whitespace__at____all"""
        output = re.sub(
            multiple_whitespace_regex,
            multiple_whitespace_replacement,
            input_string
        )
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output)
        )


class EvaluateOuterWhitespaceReplacement(TestCase):
    regex_name = 'outer_whitespace_regex'

    def test_leading_and_trailing_space(self):
        input_string = '   leading and trailing    \n\n'
        expected_output = 'leading and trailing'
        output = re.sub(
            outer_whitespace_regex,
            outer_whitespace_replacement,
            input_string
        )
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output)
        )

    def test_punctuation_before_trailing(self):
        input_string = 'only trailing space!   \t'
        expected_output = 'only trailing space!'
        output = re.sub(
            outer_whitespace_regex,
            outer_whitespace_replacement,
            input_string
        )
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output)
        )

    def test_punctuation_before_leading(self):
        input_string = '\n"Get to the back of the ship!" Tom said sternly.\n'
        expected_output = '"Get to the back of the ship!" Tom said sternly.'
        output = re.sub(
            outer_whitespace_regex,
            outer_whitespace_replacement,
            input_string
        )
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output)
        )

    def test_no_outer_space(self):
        input_string = 'Do not launch me out of the atmosphere please!'
        expected_output = 'Do not launch me out of the atmosphere please!'
        output = re.sub(
            outer_whitespace_regex,
            outer_whitespace_replacement,
            input_string
        )
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output)
        )


class EvaluateNameSwitchReplacement(TestCase):
    regex_name = 'name_switch_regex'

    def test_instructions_example(self):
        input_string = """Indira Ghandi
Morgan Freeman
Justin Finch-Fletchley
Flannery O'Connor"""
        expected_output = """Ghandi, Indira
Freeman, Morgan
Finch-Fletchley, Justin
O'Connor, Flannery"""
        output = re.sub(name_switch_regex, name_switch_replacement, input_string)
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output)
        )

    def test_single_name_string(self):
        input_string = 'Indira Ghandi'
        expected_output = 'Ghandi, Indira'
        output = re.sub(name_switch_regex, name_switch_replacement, input_string)
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output))

    def test_single_word_name(self):
        input_string = 'Pele'
        expected_output = 'Pele'
        output = re.sub(name_switch_regex, name_switch_replacement, input_string)
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output))


class EvaluateMultiNameSwitchReplacement(TestCase):
    regex_name = 'multi_name_switch_regex'

    def test_instructions_example(self):
        input_string = """Indira Ghandi
Morgan Freeman
Justin Finch-Fletchley
Flannery O'Connor
Carlos Alberto Torres
Harriet Beecher Stowe"""
        expected_output = """Ghandi, Indira
Freeman, Morgan
Finch-Fletchley, Justin
O'Connor, Flannery
Torres, Carlos Alberto
Stowe, Harriet Beecher"""
        output = re.sub(
            multi_name_switch_regex,
            multi_name_switch_replacement,
            input_string
        )
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output))

    def test_single_name(self):
        input_string = 'Indira Ghandi'
        expected_output = 'Ghandi, Indira'
        output = re.sub(
            multi_name_switch_regex,
            multi_name_switch_replacement,
            input_string
        )
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output))

    def test_single_word_string(self):
        input_string = 'Pele'
        expected_output = 'Pele'
        output = re.sub(
            multi_name_switch_regex,
            multi_name_switch_replacement,
            input_string
        )
        self.assertEqual(
            output,
            expected_output,
            output_does_not_match(self.regex_name, input_string, expected_output, output))
